package com.trendmicro.mist.proto;

option java_package = "com.trendmicro.mist.proto";
option java_outer_classname = "GateTalk";

message Connection
{
    required string host_name = 1;
    optional string host_port = 2 [ default = "" ];
    required string username = 3;
    required string password = 4;
    required string broker_type = 5 [ default = "" ];
}

message Channel
{
    enum Type 
    {
        QUEUE = 0;
        TOPIC = 1;
    }
    required string name = 1;
    required Type type = 2;
    optional bool persistent = 3 [ default = false ];
}

message Session
{
    required Connection connection = 1;
}

message Client
{
    enum Type 
    {
        CONSUMER = 0;
        PRODUCER = 1;
    }
    enum Action
    {
        MOUNT = 0;
        UNMOUNT = 1;
    }
    required uint32 session_id = 1;
    required Channel channel = 2;
    required Type type = 3;
    required Action action = 4;
}

message Request
{
    enum Type 
    {
        SESSION_LIST = 0;
        SESSION_DESTROY = 1;
        SESSION_CLEAN_FREE = 2;
        CLIENT_ATTACH = 3;
        CLIENT_DETACH = 4;
        DAEMON_STATUS = 5;
        SESSION_INFO = 6;
        PING = 7;
    }
    enum Role 
    {
        SESSION = 0;
        SINK = 1;
        SOURCE = 2;
    }
    required Type type = 1;
    optional string argument = 2 [ default = "" ];
    optional Role role = 3;
}

message Response 
{
    required bool success = 1;
    optional string context = 2 [ default = "" ];
    optional string exception = 3 [ default = "" ];
}

message Command
{
    repeated Session session = 1;
    repeated Client client = 2;
    repeated Request request = 3;
    repeated Response response = 4;
}
